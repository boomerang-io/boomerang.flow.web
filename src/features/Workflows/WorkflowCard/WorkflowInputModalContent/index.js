import React from "react";
import PropTypes from "prop-types";
import { Button, ModalBody, ModalFooter } from "carbon-components-react";
import { DynamicFormik, ModalFlowForm } from "@boomerang/carbon-addons-boomerang-react";
import ValidateFormikOnMount from "Components/ValidateFormikOnMount";
import styles from "./workflowInputModalContent.module.scss";

WorkflowInputModalContent.propTypes = {
  closeModal: PropTypes.func,
  executeWorkflow: PropTypes.func.isRequired,
  inputs: PropTypes.array.isRequired
};

function WorkflowInputModalContent({ closeModal, executeWorkflow, inputs }) {
  return (
    <DynamicFormik
      validateOnMount
      inputs={inputs}
      toggleProps={() => ({
        orientation: "vertical"
      })}
      onSubmit={(values) => { 
        const redirect = values.redirect;
        delete values.redirect;
        executeWorkflow({
          redirect,
          properties: values
        });
        closeModal();
        }
      }
    >
      {({ inputs, formikProps }) => (
        <ModalFlowForm className={styles.container}>
          <ModalBody>{inputs}</ModalBody>
          <ModalFooter>
            <Button kind="secondary" onClick={closeModal} type="button">
              Cancel
            </Button>
            <Button
              disabled={!formikProps.isValid}
              p
              onClick={e => {
                formikProps.setFieldValue("redirect", false);
                formikProps.handleSubmit();
              }}
              type="button"
            >
              Run
            </Button>
            <Button
              disabled={!formikProps.isValid}
              onClick={e => {
                formikProps.setFieldValue("redirect", true);
                formikProps.handleSubmit();
              }}
              type="button"
            >
              Run and View
            </Button>
          </ModalFooter>
          <ValidateFormikOnMount validateForm={formikProps.validateForm} />
        </ModalFlowForm>
      )}
    </DynamicFormik>
  );
}

export default WorkflowInputModalContent;
