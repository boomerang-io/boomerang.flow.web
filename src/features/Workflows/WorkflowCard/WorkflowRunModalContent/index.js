import React from "react";
import PropTypes from "prop-types";
import {
  Button,
  InlineNotification,
  ModalForm,
  ModalBody,
  ModalFooter,
} from "@boomerang/carbon-addons-boomerang-react";

WorkflowRunModalContent.propTypes = {
  closeModal: PropTypes.func.isRequired,
  executeError: PropTypes.object,
  executeWorkflow: PropTypes.func.isRequired,
  isExecuting: PropTypes.bool.isRequired,
};

function WorkflowRunModalContent({ closeModal, executeError, executeWorkflow, isExecuting }) {
  return (
    <ModalForm>
      {executeError && (
        <ModalBody>
          <InlineNotification kind="error" title="Something's Wrong" subtitle="Request to execute workflow failed" />
        </ModalBody>
      )}
      <ModalFooter>
        <Button kind="secondary" type="button" onClick={closeModal}>
          Cancel
        </Button>
        {isExecuting ? (
          <Button disabled style={{ flex: "0 1 115%" }}>
            Running...
          </Button>
        ) : (
          <>
            <Button
              onClick={(e) => {
                e.preventDefault();
                executeWorkflow({
                  closeModal,
                  redirect: false,
                });
              }}
            >
              Run
            </Button>
            <Button
              onClick={(e) => {
                e.preventDefault();
                executeWorkflow({
                  closeModal,
                  redirect: true,
                });
              }}
            >
              Run and View
            </Button>
          </>
        )}
      </ModalFooter>
    </ModalForm>
  );
}

export default WorkflowRunModalContent;
