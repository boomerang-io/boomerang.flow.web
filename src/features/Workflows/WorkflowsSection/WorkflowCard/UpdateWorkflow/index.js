import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import queryString from "query-string";
import { actions as importWorkflowActions } from "State/importWorkflow";
import { notify, ToastNotification, ModalFlow } from "@boomerang/carbon-addons-boomerang-react";
import ImportWorkflowContent from "./ImportWorkflowContent";
import { BASE_SERVICE_URL } from "Config/servicesConfig";
import styles from "./updateWorkflow.module.scss";

class UpdateWorkflow extends Component {
  static propTypes = {
    fetchTeams: PropTypes.func.isRequired,
    importWorkflowActions: PropTypes.object.isRequired,
    importWorkflowState: PropTypes.object.isRequired,
    onCloseModal: PropTypes.func,
    workflowId: PropTypes.string.isRequired
  };

  handleImportWorkflow = (data, closeModal) => {
    const query = queryString.stringify({ update: true, flowTeamId: this.props.teamId });
    return this.props.importWorkflowActions
      .post(`${BASE_SERVICE_URL}/workflow/import?${query}`, data)
      .then(() => {
        notify(<ToastNotification kind="success" title="Update Workflow" subtitle="Workflow successfullyupdated" />);
        closeModal();
        this.props.fetchTeams();
      })
      .catch(err => {
        //noop
      });
  };

  render() {
    const initialState = {
      step: 0,
      formData: {
        files: []
      }
    };
    const { isPosting, workflowId } = this.props;
    return (
      <ModalFlow
        isOpen
        confirmModalProps={{
          title: "Are you sure?",
          children: "Your request will not be saved"
        }}
        composedModalProps={{
          containerClassName: styles.container
        }}
        modalHeaderProps={{
          title: "Update .json file"
        }}
        initialState={initialState}
        onCloseModal={this.props.onCloseModal}
      >
        <ImportWorkflowContent
          confirmButtonText={isPosting ? "Updating..." : "Update"}
          handleImportWorkflow={this.handleImportWorkflow}
          isLoading={isPosting}
          title="Update a Workflow - Select the Workflow file you want to upload"
          workflowId={workflowId}
        />
      </ModalFlow>
    );
  }
}

const mapStateToProps = state => ({
  importWorkflowState: state.importWorkflow,
  isPosting: state.importWorkflow.isPosting
});

const mapDispatchToProps = dispatch => ({
  importWorkflowActions: bindActionCreators(importWorkflowActions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(UpdateWorkflow);
