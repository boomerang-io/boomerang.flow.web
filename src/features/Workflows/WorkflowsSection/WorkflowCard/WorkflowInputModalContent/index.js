import React from "react";
import PropTypes from "prop-types";
import { Button, ModalBody, ModalFooter } from "carbon-components-react";
import { DynamicFormik, ModalFlowForm } from "@boomerang/carbon-addons-boomerang-react";
import ValidateFormikOnMount from "Components/ValidateFormikOnMount";
import styles from "./workflowInputModalContent.module.scss";

WorkflowInputModalContent.propTypes = {
  closeModal: PropTypes.func,
  executeWorkflow: PropTypes.func.isRequired,
  inputs: PropTypes.array.isRequired
};

function WorkflowInputModalContent({ closeModal, executeWorkflow, inputs }) {
  return (
    <DynamicFormik
      validateOnMount
      inputs={inputs}
      dataDrivenInputProps={{
        description: null
      }}
      toggleProps={() => ({
        orientation: "vertical"
      })}
    >
      {({ inputs, formikProps }) => (
        <ModalFlowForm className={styles.container}>
          <ModalBody>{inputs}</ModalBody>
          <ModalFooter>
            <Button kind="secondary" onClick={closeModal} type="button">
              Cancel
            </Button>
            <Button
              disabled={!formikProps.isValid}
              p
              onClick={e => {
                e.preventDefault();
                executeWorkflow({
                  redirect: false,
                  properties: formikProps.values
                });
                closeModal();
              }}
              type="button"
            >
              Run
            </Button>
            <Button
              disabled={!formikProps.isValid}
              onClick={e => {
                e.preventDefault();
                executeWorkflow({
                  redirect: true,
                  properties: formikProps.values
                });
                closeModal();
              }}
              type="button"
            >
              Run and View
            </Button>
          </ModalFooter>
          <ValidateFormikOnMount validateForm={formikProps.validateForm} />
        </ModalFlowForm>
      )}
    </DynamicFormik>
  );
}

export default WorkflowInputModalContent;
