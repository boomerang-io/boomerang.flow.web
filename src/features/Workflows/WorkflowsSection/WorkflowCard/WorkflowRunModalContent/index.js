import React from "react";
import PropTypes from "prop-types";
import { Button, ModalFooter } from "carbon-components-react";
import { ModalFlowForm } from "@boomerang/carbon-addons-boomerang-react";

WorkflowRunModalContent.propTypes = {
  closeModal: PropTypes.func,
  executeWorkflow: PropTypes.func.isRequired
};

function WorkflowRunModalContent({ closeModal, executeWorkflow }) {
  return (
    <ModalFlowForm>
      <ModalFooter>
        <Button kind="secondary" type="button" onClick={closeModal}>
          Cancel
        </Button>
        <Button
          onClick={e => {
            e.preventDefault();
            executeWorkflow({
              redirect: false
            });
            closeModal();
          }}
        >
          Run
        </Button>
        <Button
          onClick={e => {
            e.preventDefault();
            executeWorkflow({
              redirect: true
            });
            closeModal();
          }}
        >
          Run and View
        </Button>
      </ModalFooter>
    </ModalFlowForm>
  );
}

export default WorkflowRunModalContent;
