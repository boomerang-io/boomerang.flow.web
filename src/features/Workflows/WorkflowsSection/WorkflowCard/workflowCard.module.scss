@import "~@boomerang/boomerang-styles/lib/colors";
@import "~Styles/mixins";

.container {
  display: flex;
  background-color: $white;
  flex-direction: column;
  justify-content: space-between;
  min-width: 19.375rem;
  width: 19.375rem;
  margin: 1rem 0.75rem;
  height: 10.31255rem;
  position: relative;
  overflow: hidden;
  @include border-radius();

  :global .bx--overflow-menu {
    height: 2rem;
  }

  :global .bx--overflow-menu-options__btn {
    color: $gray-50;
    text-decoration: none !important;
  }
  :global .bx--overflow-menu-options--open {
    @include border-radius();
    overflow: hidden;
    z-index: 998;
  }

  :global .bx--overflow-menu-options__option {
    min-width: 11.125rem;
    // &:hover {
    //   background-color: $cool-gray-1;
    // }
  }
}
.cardHeader {
  position: absolute;
  right: 0;
}
.cardStatusContainer {
  margin-left: 0.75rem;
  display: flex;
  align-items: center;
}
.cardStatus {
  color: $gray-50;
}
.cardCircle {
  margin-right: 0.5rem;
  min-width: 0.75rem;
  max-width: 0.75rem;
  min-height: 0.75rem;
  max-height: 0.75rem;
  border-radius: 50%;
  &.--published {
    background-color: $green-30;
  }
  &.--draft {
    background-color: $gray-20;
  }
}

.cardInfo {
  display: flex;
  justify-content: space-between;
  height: 8rem;
  padding: 0.5rem;
}
.cardIconContainer {
  align-items: center;
  display: flex;
  justify-content: center;
  width: 5rem;
}
.cardIcon {
  width: 2rem;
  height: 2rem;
  margin-bottom: 2.5rem;
}
.cardDescriptionContainer {
  color: $gray-60;
  cursor: pointer;
  display: flex;
  font-size: 0.875rem;
  flex-direction: column;
  margin: 1rem 1rem 0rem 0;
  overflow: hidden;
  width: 15rem;
  &:focus {
    outline: none;
  }
}
.cardName {
  color: #242a2e;
  font-size: 1.125rem;
  font-weight: 600;
  letter-spacing: 0.0625rem;
  line-height: 1.25rem;
  margin-bottom: 0.5rem;
  text-align: left;
}

.cardDescription {
  color: #242a2e;
  display: block;
  font-size: 0.75rem;
  letter-spacing: 0.0625rem;
  line-height: 0.92rem;
  text-align: left;
  text-overflow: ellipsis;
  word-break: break-word;
}

.cardLaunch {
  cursor: pointer;
  width: 100%;
  :global .bx--btn {
    border-top-left-radius: 0;
    border-top-right-radius: 0;
    flex-direction: row-reverse;
    font-size: 1rem;
    font-weight: 600;
    justify-content: center;
    max-width: 19.375rem;
    padding: 0rem 0.25rem;
    width: 19.375rem;

    .bx--btn__icon {
      height: 1.5rem;
      margin-right: 1rem;
      position: initial;
      width: 1.5rem;
    }
  }
}
.executeWorkflow {
  height: 16rem;
  width: 30rem;
  text-align: left;
}
:global .bx--overflow-menu-options.bx--overflow-menu--flip {
  position: fixed !important;
}
