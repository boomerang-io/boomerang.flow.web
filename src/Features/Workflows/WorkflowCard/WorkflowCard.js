import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { useMutation, queryCache } from "react-query";
import { Link, useHistory } from "react-router-dom";
import {
  Button,
  ConfirmModal,
  ComposedModal,
  InlineLoading,
  OverflowMenu,
  OverflowMenuItem,
  ToastNotification,
  TooltipIcon,
  notify,
} from "@boomerang-io/carbon-addons-boomerang-react";
import WorkflowWarningButton from "Components/WorkflowWarningButton";
import UpdateWorkflow from "./UpdateWorkflow";
import WorkflowInputModalContent from "./WorkflowInputModalContent";
import WorkflowRunModalContent from "./WorkflowRunModalContent";
import fileDownload from "js-file-download";
import { appLink } from "Config/appConfig";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { BASE_SERVICE_URL } from "Config/servicesConfig";
import { Run20, Bee20 } from "@carbon/icons-react";
import workflowIcons from "Assets/workflowIcons";
import styles from "./workflowCard.module.scss";

WorkflowCard.propTypes = {
  teamId: PropTypes.string.isRequired,
  workflow: PropTypes.object.isRequired,
};

function WorkflowCard({ teamId, workflow }) {
  const cancelRequestRef = React.useRef();
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
  const [isUpdateWorkflowModalOpen, setIsUpdateWorkflowModalOpen] = useState(false);

  const history = useHistory();

  const [deleteWorkflowMutator, { isLoading: isDeleting }] = useMutation(resolver.deleteWorkflow, {
    onSuccess: () => queryCache.invalidateQueries(serviceUrl.getTeams()),
  });

  const [executeWorkflowMutator, { error: executeError, isLoading: isExecuting }] = useMutation((args) => {
    const { promise, cancel } = resolver.postExecuteWorkflow(args);
    cancelRequestRef.current = cancel;
    return promise;
  });

  const handleOverflowMenuOpen = () => {
    window.addEventListener("keydown", preventKeyScrolling, false);
  };

  const handleOverflowMenuClose = () => {
    window.removeEventListener("keydown", preventKeyScrolling, false);
  };

  useEffect(() => {
    return handleOverflowMenuClose();
  });

  /**
   * Format properties to be edited in form by Formik. It doesn't work with property notation :(
   * See: https://jaredpalmer.com/formik/docs/guides/arrays#nested-objects
   * This is safe to do because we don't accept "-" characters in property keys
   * @returns {Array}
   */
  const formatPropertiesForEdit = () => {
    const { properties = [] } = workflow;
    return properties.filter((property) => !property.readOnly);
  };

  const handleDeleteWorkflow = async () => {
    try {
      await deleteWorkflowMutator({ id: workflow.id });
      notify(<ToastNotification kind="success" title="Delete Workflow" subtitle="Workflow successfully deleted" />);
    } catch {
      notify(<ToastNotification kind="error" title="Something's Wrong" subtitle="Request to delete workflow failed" />);
    }
  };

  const handleExportWorkflow = (workflow) => {
    notify(<ToastNotification kind="info" title="Export Workflow" subtitle="Export starting soon" />);
    axios
      .get(`${BASE_SERVICE_URL}/workflow/export/${workflow.id}`)
      .then(({ data }) => {
        fileDownload(JSON.stringify(data, null, 4), `${workflow.name}.json`);
      })
      .catch((error) => {
        notify(<ToastNotification kind="error" title="Something's Wrong" subtitle="Export workflow failed" />);
      });
  };

  const handleExecuteWorkflow = async ({ closeModal, redirect = false, properties = {} }) => {
    const { id: workflowId } = workflow;
    try {
      const { data: execution } = await executeWorkflowMutator({ id: workflowId, properties });
      notify(
        <ToastNotification kind="success" title="Run Workflow" subtitle="Successfully started workflow execution" />
      );
      if (redirect) {
        history.push({
          pathname: appLink.execution({ executionId: execution.id, workflowId }),
          state: { fromUrl: appLink.workflows(), fromText: "Workflows" },
        });
      } else {
        closeModal();
      }
    } catch {
      //no-op
    }
  };

  /* prevent page scroll when up or down arrows are pressed **/
  const preventKeyScrolling = (e) => {
    if ([38, 40].indexOf(e.keyCode) > -1) {
      e.preventDefault();
    }
  };

  const menuOptions = [
    {
      itemText: "Edit Workflow",
      onClick: () => history.push(appLink.editorDesigner({ teamId: workflow.flowTeamId, workflowId: workflow.id })),
      primaryFocus: true,
    },
    {
      itemText: "View Activity",
      onClick: () => history.push(appLink.workflowActivity({ workflowId: workflow.id })),
    },

    {
      itemText: "Export .json",
      onClick: () => handleExportWorkflow(workflow),
    },
    {
      itemText: "Update .json",
      onClick: () => setIsUpdateWorkflowModalOpen(true),
    },
    {
      hasDivider: true,
      itemText: "Delete",
      isDelete: true,
      onClick: () => setIsDeleteModalOpen(true),
    },
  ];

  const formattedProperties = formatPropertiesForEdit();

  const { name, Icon = Bee20 } = workflowIcons.find((icon) => icon.name === workflow.icon) ?? {};

  return (
    <div className={styles.container}>
      <Link disabled={isDeleting} to={appLink.editorDesigner({ teamId: workflow.flowTeamId, workflowId: workflow.id })}>
        <section className={styles.details}>
          <div className={styles.iconContainer}>
            <Icon className={styles.icon} alt={`${name}`} />
          </div>
          <div className={styles.descriptionContainer}>
            <h1 title={workflow.name} className={styles.name} data-testid="workflow-card-title">
              {workflow.name}
            </h1>
            <p title={workflow.shortDescription} className={styles.description}>
              {workflow.shortDescription}
            </p>
          </div>
        </section>
      </Link>
      <section className={styles.launch}>
        {Array.isArray(formattedProperties) && formattedProperties.length !== 0 ? (
          <ComposedModal
            modalHeaderProps={{
              title: "Workflow Properties",
              subtitle: "Provide property values for your workflow",
            }}
            modalTrigger={({ openModal }) => (
              <Button
                disabled={isDeleting}
                iconDescription="Run Workflow"
                renderIcon={Run20}
                size="field"
                onClick={openModal}
              >
                Run it
              </Button>
            )}
            onCloseModal={() => {
              if (cancelRequestRef.current) cancelRequestRef.current();
            }}
          >
            {({ closeModal }) => (
              <WorkflowInputModalContent
                closeModal={closeModal}
                executeError={executeError}
                executeWorkflow={handleExecuteWorkflow}
                isExecuting={isExecuting}
                inputs={formattedProperties}
              />
            )}
          </ComposedModal>
        ) : (
            <ComposedModal
              composedModalProps={{ containerClassName: `${styles.executeWorkflow}` }}
              modalHeaderProps={{
                title: "Execute Workflow",
                subtitle: '"Run and View" will navigate you to the workflow exeuction view.',
              }}
              modalTrigger={({ openModal }) => (
                <Button
                  disabled={isDeleting}
                  iconDescription="Run Workflow"
                  renderIcon={Run20}
                  size="field"
                  onClick={openModal}
                >
                  Run it
              </Button>
              )}
            >
              {({ closeModal }) => (
                <WorkflowRunModalContent
                  closeModal={closeModal}
                  executeError={executeError}
                  executeWorkflow={handleExecuteWorkflow}
                  isExecuting={isExecuting}
                />
              )}
            </ComposedModal>
          )}
      </section>
      {workflow.templateUpgradesAvailable && (
        <div className={styles.templatesWarningIcon}>
          <TooltipIcon direction="top" tooltipText={"New version of a task available! To update, edit your workflow."}>
            <WorkflowWarningButton />
          </TooltipIcon>
        </div>
      )}
      {isDeleting || isExecuting ? (
        <InlineLoading
          description={isDeleting ? "Deleting..." : "Executing..."}
          style={{ position: "absolute", right: "0.5rem", top: "0", width: "fit-content" }}
        />
      ) : (
          <OverflowMenu
            flipped
            ariaLabel="Overflow card menu"
            iconDescription="Overflow menu icon"
            onOpen={handleOverflowMenuOpen}
            onClose={handleOverflowMenuClose}
            style={{ position: "absolute", right: "0" }}
          >
            {menuOptions.map(({ onClick, itemText, ...rest }, index) => (
              <OverflowMenuItem onClick={onClick} itemText={itemText} key={`${itemText}-${index}`} {...rest} />
            ))}
          </OverflowMenu>
        )}
      {isUpdateWorkflowModalOpen && (
        <UpdateWorkflow
          onCloseModal={() => setIsUpdateWorkflowModalOpen(false)}
          teamId={teamId}
          workflowId={workflow.id}
        />
      )}
      {isDeleteModalOpen && (
        <ConfirmModal
          affirmativeAction={handleDeleteWorkflow}
          affirmativeButtonProps={{ kind: "danger" }}
          affirmativeText="Delete"
          isOpen={isDeleteModalOpen}
          negativeAction={() => {
            setIsDeleteModalOpen(false);
          }}
          negativeText="Cancel"
          onCloseModal={() => {
            setIsDeleteModalOpen(false);
          }}
          title="Delete Workflow"
        >
          Are you sure you want to delete this workflow? There's no going back from this Decision.
        </ConfirmModal>
      )}
    </div>
  );
}

export default WorkflowCard;
