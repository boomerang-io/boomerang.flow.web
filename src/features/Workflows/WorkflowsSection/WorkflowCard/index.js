import React, { Component } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { Link, withRouter } from "react-router-dom";
import fileDownload from "js-file-download";
import { Button, OverflowMenu, OverflowMenuItem } from "carbon-components-react";
import {
  notify,
  ToastNotification,
  ModalFlow,
  ConfirmModal,
  TooltipIcon,
} from "@boomerang/carbon-addons-boomerang-react";
import WorkflowWarningButton from "Components/WorkflowWarningButton";
import WorkflowInputModalContent from "./WorkflowInputModalContent";
import WorkflowRunModalContent from "./WorkflowRunModalContent";
import UpdateWorkflow from "./UpdateWorkflow";
import imgs from "Assets/icons";
import { Run20 } from "@carbon/icons-react";
import { BASE_SERVICE_URL } from "Config/servicesConfig";
import styles from "./workflowCard.module.scss";

class WorkflowCard extends Component {
  static propTypes = {
    deleteWorkflow: PropTypes.func.isRequired,
    executeWorkflow: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    setActiveTeam: PropTypes.func.isRequired,
    teamId: PropTypes.string.isRequired,
    workflow: PropTypes.object.isRequired,
  };

  state = {
    isDeleteModalOpen: false,
    isUpdateWorkflowModalOpen: false,
  };

  componentWillUnmount() {
    this.handleOverflowMenuClose();
  }

  executeWorkflow = ({ redirect, properties }) => {
    this.props.executeWorkflow({
      workflowId: this.props.workflow.id,
      redirect,
      properties,
    });
  };

  /**
   * Format properties to be edited in form by Formik. It doesn't work with property notation :(
   * See: https://jaredpalmer.com/formik/docs/guides/arrays#nested-objects
   * This is safe to do because we don't accept "-" characters in property keys
   * @returns {Array}
   */
  formatPropertiesForEdit = () => {
    const { properties = [] } = this.props.workflow;
    return properties.filter((property) => !property.readOnly);
  };

  handleExportWorkflow = (workflow) => {
    notify(<ToastNotification kind="info" title="Export Workflow" subtitle="Your download will start soon." />);
    axios
      .get(`${BASE_SERVICE_URL}/workflow/export/${workflow.id}`)
      .then((res) => {
        const status = res.status.toString();
        if (status.startsWith("4") || status.startsWith("5"))
          notify(<ToastNotification kind="error" title="Export Workflow" subtitle="Something went wrong." />);
        else fileDownload(JSON.stringify(res.data, null, 4), `${workflow.name}.json`);
      })
      .catch((error) => {
        notify(<ToastNotification kind="error" title="Export Workflow" subtitle="Something went wrong." />);
      });
  };

  setActiveTeam = () => {
    const { setActiveTeam, teamId } = this.props;
    setActiveTeam(teamId);
  };

  setActiveTeamAndRedirect = () => {
    const { history, workflow } = this.props;
    this.setActiveTeam();
    history.push(`/editor/${workflow.id}/designer`);
  };

  /* prevent page scroll when up or down arrows are pressed **/
  preventKeyScrolling = (e) => {
    if ([38, 40].indexOf(e.keyCode) > -1) {
      e.preventDefault();
    }
  };

  handleOverflowMenuOpen = () => {
    window.addEventListener("keydown", this.preventKeyScrolling, false);
  };

  handleOverflowMenuClose = () => {
    window.removeEventListener("keydown", this.preventKeyScrolling, false);
  };

  render() {
    const { deleteWorkflow, fetchTeams, history, teamId, workflow } = this.props;
    const menuOptions = [
      {
        itemText: "Edit Workflow",
        onClick: this.setActiveTeamAndRedirect,
        primaryFocus: true,
      },
      {
        itemText: "View Activity",
        onClick: () => history.push(`/activity?page=0&size=10&workflowIds=${workflow.id}`),
      },

      {
        itemText: "Export .json",
        onClick: () => this.handleExportWorkflow(workflow),
      },
      {
        itemText: "Update .json",
        onClick: () => this.setState({ isUpdateWorkflowModalOpen: true }),
      },
      {
        hasDivider: true,
        itemText: "Delete",
        isDelete: true,
        onClick: () => this.setState({ isDeleteModalOpen: true }),
      },
    ];

    const formattedProperties = this.formatPropertiesForEdit();

    return (
      <div className={styles.container}>
        <Link to={`/editor/${workflow.id}/designer`} onClick={this.setActiveTeam}>
          <section className={styles.cardInfo}>
            <div className={styles.cardIconContainer}>
              <img className={styles.cardIcon} src={imgs[workflow.icon ? workflow.icon : "docs"]} alt="icon" />
            </div>
            <div className={styles.cardDescriptionContainer}>
              <h2 className={styles.cardName}>{workflow.name}</h2>
              <p className={styles.cardDescription}>{workflow.shortDescription}</p>
            </div>
          </section>
        </Link>
        <section className={styles.cardLaunch}>
          {Array.isArray(formattedProperties) && formattedProperties.length !== 0 ? (
            <ModalFlow
              modalHeaderProps={{
                title: "Workflow Properties",
                subtitle: "Provide property values for your workflow",
              }}
              modalTrigger={({ openModal }) => (
                <Button iconDescription="Run Workflow" renderIcon={Run20} size="field" onClick={openModal}>
                  Run it
                </Button>
              )}
            >
              <WorkflowInputModalContent executeWorkflow={this.executeWorkflow} inputs={formattedProperties} />
            </ModalFlow>
          ) : (
            <ModalFlow
              composedModalProps={{ containerClassName: `${styles.executeWorkflow}` }}
              modalHeaderProps={{
                title: "Execute Workflow?",
                subtitle: '"Run and View" will navigate you to the workflow exeuction view.',
              }}
              modalTrigger={({ openModal }) => (
                <Button iconDescription="Run Workflow" renderIcon={Run20} size="field" onClick={openModal}>
                  Run it
                </Button>
              )}
            >
              <WorkflowRunModalContent executeWorkflow={this.executeWorkflow} />
            </ModalFlow>
          )}
        </section>
        {workflow.templateUpgradesAvailable && (
          <div className={styles.templatesWarningIcon}>
            <TooltipIcon
              direction="top"
              tooltipText={"New version of a task available! To update, edit your workflow."}
            >
              <WorkflowWarningButton />
            </TooltipIcon>
          </div>
        )}
        <OverflowMenu
          flipped
          ariaLabel="Overflow card menu"
          iconDescription="Overflow menu icon"
          onOpen={this.handleOverflowMenuOpen}
          onClose={this.handleOverflowMenuClose}
          style={{ position: "absolute", right: "0" }}
        >
          {menuOptions.map(({ onClick, itemText, ...rest }, index) => (
            <OverflowMenuItem onClick={onClick} itemText={itemText} key={`${itemText}-${index}`} {...rest} />
          ))}
        </OverflowMenu>
        {this.state.isUpdateWorkflowModalOpen && (
          <UpdateWorkflow
            fetchTeams={fetchTeams}
            onCloseModal={() => this.setState({ isUpdateWorkflowModalOpen: false })}
            teamId={teamId}
            workflowId={workflow.id}
          />
        )}
        {this.state.isDeleteModalOpen && (
          <ConfirmModal
            affirmativeAction={() => {
              deleteWorkflow({ teamId, workflowId: workflow.id });
            }}
            affirmativeButtonProps={{ kind: "danger" }}
            affirmativeText="Delete"
            isOpen={this.state.isDeleteModalOpen}
            negativeAction={() => {
              this.setState({ isDeleteModalOpen: false });
            }}
            negativeText="Cancel"
            onCloseModal={() => {
              this.setState({ isDeleteModalOpen: false });
            }}
            title="Delete Workflow"
          >
            Are you sure you want to delete this workflow? There's no going back from this decision.
          </ConfirmModal>
        )}
      </div>
    );
  }
}

export default withRouter(WorkflowCard);
