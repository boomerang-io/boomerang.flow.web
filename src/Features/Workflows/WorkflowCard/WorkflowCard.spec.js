import React from "react";
import WorkflowCard from "./index";
import { queryCaches } from "react-query";

const mockfn = jest.fn();

const props = {
  history: {},
  teamId: "1",
  updateWorkflows: mockfn,
  executeWorkflow: mockfn,
  deleteWorkflow: mockfn,
  setActiveTeam: mockfn,
  workflow: {
    id: "789",
    name: "lucas-workflow-3",
    description: "blablabla",
    status: "draft",
    icon: "schedule",
  },
};

afterEach(() => {
  queryCaches.forEach((queryCache) => queryCache.clear());
});

describe("WorkflowCard --- Snapshot", () => {
  it("Capturing Snapshot of WorkflowCard", () => {
    const { baseElement } = rtlRouterRender(<WorkflowCard {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
