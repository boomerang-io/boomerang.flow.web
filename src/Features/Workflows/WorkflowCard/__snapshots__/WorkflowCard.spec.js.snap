// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowCard --- Snapshot Capturing Snapshot of WorkflowCard 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="container"
    >
      <a
        href="/teams/undefined/editor/789/workflow"
      >
        <section
          class="details"
        >
          <div
            class="iconContainer"
          >
            <svg
              alt="schedule"
              aria-hidden="true"
              class="icon"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21,30a8,8,0,1,1,8-8A8,8,0,0,1,21,30Zm0-14a6,6,0,1,0,6,6A6,6,0,0,0,21,16Z"
              />
              <path
                d="M22.59 25L20 22.41 20 18 22 18 22 21.59 24 23.59 22.59 25z"
              />
              <path
                d="M28,6a2,2,0,0,0-2-2H22V2H20V4H12V2H10V4H6A2,2,0,0,0,4,6V26a2,2,0,0,0,2,2h4V26H6V6h4V8h2V6h8V8h2V6h4v6h2Z"
              />
            </svg>
          </div>
          <div
            class="descriptionContainer"
          >
            <h1
              class="name"
              data-testid="workflow-card-title"
              title="lucas-workflow-3"
            >
              lucas-workflow-3
            </h1>
            <p
              class="description"
            />
          </div>
        </section>
      </a>
      <section
        class="launch"
      >
        <button
          class="bx--btn bx--btn--field bx--btn--primary"
          tabindex="0"
          type="button"
        >
          Run it
          <svg
            aria-hidden="true"
            aria-label="Run Workflow"
            class="bx--btn__icon"
            fill="currentColor"
            focusable="false"
            height="20"
            preserveAspectRatio="xMidYMid meet"
            role="img"
            viewBox="0 0 32 32"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21,16a6,6,0,1,1-6,6,6,6,0,0,1,6-6m0-2a8,8,0,1,0,8,8,8,8,0,0,0-8-8Z"
            />
            <path
              d="M26,4H6A2,2,0,0,0,4,6V26a2,2,0,0,0,2,2h4V26H6V12H28V6A2,2,0,0,0,26,4ZM6,10V6H26v4Z"
            />
            <path
              d="M19 19L19 25 24 22 19 19z"
            />
          </svg>
        </button>
      </section>
      <button
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="Overflow card menu"
        class="bx--overflow-menu"
        style="position: absolute; right: 0px;"
        tabindex="0"
      >
        <svg
          aria-label="Overflow menu icon"
          class="bx--overflow-menu__icon"
          fill="currentColor"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          role="img"
          viewBox="0 0 32 32"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="16"
            cy="8"
            r="2"
          />
          <circle
            cx="16"
            cy="16"
            r="2"
          />
          <circle
            cx="16"
            cy="24"
            r="2"
          />
          <title>
            Overflow menu icon
          </title>
        </svg>
      </button>
    </div>
  </div>
  <div
    class="bx--bmrg-modal-portal"
  />
</body>
`;
